src/daidepp/keywords/base_keywords.py

```diff
@@ -1,33 +1,50 @@
 from __future__ import annotations
 
-from dataclasses import dataclass
+from dataclasses import dataclass, field
 from typing import List, Optional, Union
 
+from typing_extensions import get_args
+
 from daidepp.constants import *
 from daidepp.keywords.daide_object import _DAIDEObject
 
+_prov_no_coast = [prov for lit in get_args(ProvinceNoCoast) for prov in get_args(lit)]
+
 
 @dataclass
 class Location:
-    province: Union[str, Location]
-    coast: Optional[str] = None
+    province: ProvinceNoCoast
+    coast: Optional[Coast] = None
 
-    def __post_init__(self):
-        if isinstance(self.province, Location):
-            self.province = self.province.province
+    def __init__(
+        self, province: Union[ProvinceNoCoast, Location], coast: Optional[Coast] = None
+    ) -> None:
+        if isinstance(province, Location):
+            self.province = province.province
+        else:
+            self.province = province
+        self.coast = coast
 
     def __str__(self) -> str:
         if self.coast:
             return f"({self.province} {self.coast})"
         return self.province
 
+    def __hash__(self) -> int:
```

Comment:
I think in the `@dataclass` decorator you can control the auto-implementation of a hash function.

Reply:
True, but there are complicated rules about when it will do that and I'd rather just write this method as opposed to passing `unsafe_hash=True`.